### PR TITLE
ARROW-11782: [GLib][Ruby][Dataset] Remove bindings for internal classes

### PR DESCRIPTION
**Pull request by kou**

### CODE REVIEW

File: cpp/src/arrow/dataset/discovery.h [hunk 1]
```diff
@@ -237,16 +237,22 @@ class ARROW_DS_EXPORT FileSystemDatasetFactory : public DatasetFactory {
...
   Result<std::vector<std::shared_ptr<Schema>>> InspectSchemas(
       InspectOptions options) override;
 
   Result<std::shared_ptr<Dataset>> Finish(FinishOptions options) override;
 
  protected:
-  static Result<std::shared_ptr<DatasetFactory>> Make(
-      std::shared_ptr<fs::FileSystem> filesystem, const std::vector<fs::FileInfo>& files,
-      std::shared_ptr<FileFormat> format, FileSystemFactoryOptions options);
-
```
Comment (kou) [inline cpp/src/arrow/dataset/discovery.h:249]: @bkietz Can we export this constructor? I want to use this in Apache Arrow Dataset GLib.

Comment (bkietz) [reply cpp/src/arrow/dataset/discovery.h:249]: If it's convenient for you I think it's fine

Comment (kou) [reply cpp/src/arrow/dataset/discovery.h:249]: Thanks!

**Comment by kou** [comment]
+1

**Pull request closed without merging**